File: backend/app/models/appointment.py
import enum
from sqlalchemy import Column, String, Date, Time, Text, ForeignKey, Enum as SAEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.database import Base
from app.models.mixins import TimestampMixin, UUIDMixin


class AppointmentType(str, enum.Enum):
    first_visit = "first_visit"
    follow_up = "follow_up"
    recurring = "recurring"
    walk_in = "walk_in"


class AppointmentStatus(str, enum.Enum):
    pending = "pending"
    confirmed = "confirmed"
    completed = "completed"
    cancelled = "cancelled"
    no_show = "no_show"
    rescheduled = "rescheduled"


class BookedBy(str, enum.Enum):
    patient = "patient"
    admin = "admin"


class Appointment(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "appointments"

    clinic_id = Column(UUID(as_uuid=True), ForeignKey("clinics.id"), nullable=False)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    doctor_id = Column(UUID(as_uuid=True), ForeignKey("doctors.id"), nullable=False)
    series_id = Column(UUID(as_uuid=True), ForeignKey("appointment_series.id"), nullable=True)

    appointment_type = Column(SAEnum(AppointmentType), nullable=False, default=AppointmentType.first_visit)
    date = Column(Date, nullable=False)
    start_time = Column(Time, nullable=False)
    end_time = Column(Time, nullable=False)
    status = Column(SAEnum(AppointmentStatus), default=AppointmentStatus.pending, nullable=False)

    internal_notes = Column(Text, nullable=True)       # Visible to admin + doctor only
    booked_by = Column(SAEnum(BookedBy), default=BookedBy.patient, nullable=False)
    cancellation_reason = Column(Text, nullable=True)
    rescheduled_to_id = Column(UUID(as_uuid=True), ForeignKey("appointments.id"), nullable=True)

    # Relationships
    clinic = relationship("Clinic", back_populates="appointments")
    patient = relationship("User", back_populates="appointments_as_patient", foreign_keys=[patient_id])
    doctor = relationship("Doctor", back_populates="appointments", foreign_keys=[doctor_id])
    series = relationship("AppointmentSeries", back_populates="appointments")
    reminders = relationship("Reminder", back_populates="appointment", cascade="all, delete-orphan")
    rescheduled_to = relationship("Appointment", remote_side="Appointment.id", foreign_keys=[rescheduled_to_id])

File: frontend/types/appointment.ts
export type AppointmentType = 'first_visit' | 'follow_up' | 'recurring' | 'walk_in'
export type AppointmentStatus = 'pending' | 'confirmed' | 'completed' | 'cancelled' | 'no_show' | 'rescheduled'
export type RecurrenceType = 'weekly' | 'biweekly' | 'monthly'

export interface Appointment {
  id: string
  clinic_id: string
  patient_id: string
  doctor_id: string
  series_id: string | null
  appointment_type: AppointmentType
  date: string
  start_time: string
  end_time: string
  status: AppointmentStatus
  internal_notes: string | null
  booked_by: 'patient' | 'admin'
  cancellation_reason: string | null
  rescheduled_to_id: string | null
  created_at: string
}

export interface AppointmentSeries {
  id: string
  patient_id: string
  doctor_id: string
  clinic_id: string
  recurrence_type: RecurrenceType
  preferred_time: string
  day_of_week: number
  start_date: string
  end_date: string | null
  occurrences_total: number | null
  occurrences_completed: number
  status: 'active' | 'paused' | 'cancelled'
}

export interface TimeSlot { start_time: string; end_time: string }
